File: app/src/main/java/com/mytlogos/enterprise/ui/ExternalUserListFragment.java
package com.mytlogos.enterprise.ui;

import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModelProviders;
import androidx.paging.PagedList;

import com.mytlogos.enterprise.R;
import com.mytlogos.enterprise.model.ExternalUser;
import com.mytlogos.enterprise.tools.Utils;
import com.mytlogos.enterprise.viewmodel.ExternalUserViewModel;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import eu.davidea.flexibleadapter.FlexibleAdapter;
import eu.davidea.flexibleadapter.items.AbstractFlexibleItem;
import eu.davidea.flexibleadapter.items.IFlexible;

public class ExternalUserListFragment extends BaseListFragment<ExternalUser, ExternalUserViewModel> {
    @Override
    ExternalUserViewModel createViewModel() {
        return ViewModelProviders.of(this).get(ExternalUserViewModel.class);
    }

    @Override
    LiveData<PagedList<ExternalUser>> createPagedListLiveData() {
        return getViewModel().getExternalUser();
    }

    @Override
    List<IFlexible> convertToFlexibles(Collection<ExternalUser> list) {
        List<IFlexible> flexibles = new ArrayList<>();

        for (ExternalUser user : list) {
            if (user == null) {
                break;
            }
            flexibles.add(new UserItem(user));
        }
        return flexibles;
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        // TODO: 02.08.2019 implement add external user
    }

    private static class UserItem extends AbstractFlexibleItem<MetaViewHolder> {
        private final ExternalUser externalUser;

        private UserItem(@NonNull ExternalUser externalUser) {
            this.externalUser = externalUser;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            UserItem userItem = (UserItem) o;

            return externalUser.equals(userItem.externalUser);
        }

        @Override
        public int hashCode() {
            return externalUser.hashCode();
        }

        @Override
        public int getLayoutRes() {
            return R.layout.meta_item;
        }

        @Override
        public MetaViewHolder createViewHolder(View view, FlexibleAdapter<IFlexible> adapter) {
            return new MetaViewHolder(view, adapter);
        }

        @Override
        public void bindViewHolder(FlexibleAdapter<IFlexible> adapter, MetaViewHolder holder, int position, List<Object> payloads) {
            holder.topLeftText.setText(Utils.externalUserTypeToName(this.externalUser.getType()));
            holder.mainText.setText(this.externalUser.getIdentifier());
        }
    }
}

File: app/src/main/java/com/mytlogos/enterprise/ui/ExternalUserListFragment.kt
package com.mytlogos.enterprise.ui

import android.annotation.SuppressLint
import android.view.View
import androidx.paging.PagingData
import androidx.recyclerview.widget.DiffUtil
import com.mytlogos.enterprise.R
import com.mytlogos.enterprise.model.ExternalUser
import com.mytlogos.enterprise.tools.externalUserTypeToName
import com.mytlogos.enterprise.viewmodel.ExternalUserViewModel
import kotlinx.coroutines.flow.Flow

class ExternalUserListFragment : BasePagingFragment<ExternalUser, ExternalUserViewModel>() {
    override val viewModelClass: Class<ExternalUserViewModel>
        get() = ExternalUserViewModel::class.java

    override fun createAdapter(): BaseAdapter<ExternalUser, *> = UserAdapter()

    override fun createPaged(model: ExternalUserViewModel): Flow<PagingData<ExternalUser>> {
        return viewModel.externalUser
    }

    private class UserDiff : DiffUtil.ItemCallback<ExternalUser>() {
        override fun areItemsTheSame(oldItem: ExternalUser, newItem: ExternalUser): Boolean {
            return oldItem.uuid == newItem.uuid
        }

        override fun areContentsTheSame(oldItem: ExternalUser, newItem: ExternalUser): Boolean {
            return oldItem == newItem
        }
    }

    private class UserAdapter : BaseAdapter<ExternalUser, NewMetaViewHolder>(UserDiff()) {

        @SuppressLint("SetTextI18n")
        override fun onBindViewHolder(holder: NewMetaViewHolder, position: Int) {
            val item = getItem(position)

            if (item != null) {
                holder.topLeftText.text = externalUserTypeToName(item.type)
                holder.mainText.text = item.identifier
            }
        }

        override val layoutId = R.layout.meta_item

        override fun createViewHolder(root: View, viewType: Int) = NewMetaViewHolder(root)
    }

}